[Sema] Refactor member type matching from `checkOverrides` to make it reusable

xedin opened the pull request:
These changes make it possible to use the same type matching logic
for both `checkOverrides` and `addImplicitConstructors` which removes
one of the last usages of `AnyFunctionType::getInput` and correctly
handles generic types.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Sema] Refactor member type matching from `checkOverrides` to make it reusable

These changes make it possible to use the same type matching logic
for both `checkOverrides` and `addImplicitConstructors` which removes
one of the last usages of `AnyFunctionType::getInput` and correctly
handles generic types.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3996,6 +3996,125 @@ static void finalizeAbstractStorageDecl(TypeChecker &TC,
+}
+
+/// Drop the optionality of the result type of the given function type.
+static Type dropResultOptionality(Type type, unsigned uncurryLevel) {
```
Structurally, this is almost identical to `TypeBase::replaceCovariantResultType()`. Should we generalize that with some kind of "here's a function object to transform the result type" thing?

xedin replied:
Sure, I will look into that separately!

slavapestov replied:
replaceCovariantResultType() is kind of weird because it completely ignores the result type. It's used for DynamicSelf on class methods, and I think it's kind of doing the wrong thing anyway -- in its current form it can't return a class method returning (Self, Self) or anything with DynamicSelf in a non-trivial but still covariant position. I wouldn't use it as a model for anything, it should instead go away.

DougGregor reviewed (approved):
This looks good, thanks

slavapestov commented:
Can you add tests for the generic case you fixed?

xedin commented:
Absolutely! Just wanted to run the source compat tests to makes sure that I didn’t break anything.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3996,6 +3996,125 @@ static void finalizeAbstractStorageDecl(TypeChecker &TC,
+    adjustFunctionTypeForOverride(memberType);
+  }
+
+  if (subscript) {
```
Should this be an 'else if'? Otherwise you run the '// For properties...' block on methods too.

xedin replied:
Original code had a `if` in there I just moved it up, but I think you are right, it should be `else if`.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3996,6 +3996,125 @@ static void finalizeAbstractStorageDecl(TypeChecker &TC,
+  return memberType;
+}
+
+static bool isOverrideBasedOnType(ValueDecl *decl, Type declTy,
```
how about isValidTypeForOverride() or something?

xedin replied:
This method doesn't check validity, it matches type of the child to parent to see if they are the same...

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Interestingly enough it looks like cases with generic signature are caught by the decl-checking and wouldn't compile without override e.g. (Thanks to @DougGregor for help!)

```swift
class A<T> {
  init(_ foo: T) {}
}

class B<U> : A<[U]> {
  init(_ foo: [U]) {}
}
```

Today if there is no `override` it's going to be an error, and once `override` is there it's tracked by `overridenInits` that triggers another check in `addImplicitConstructors` and skips importing superclass method.

But if I make superclass one `private` it does currently fail as `re-declaration` but not with my changes which seems to be correct:

```swift
class A<T> {
  private init(_ foo: T) {}
}

class B<U> : A<[U]> {
  init(_ foo: [U]) {}
}
```

I'm going to add this last test case to the suite.

xedin committed:
[AST] `lookupSubstition` should produce nothing if there are not substitutions

If `SubstitutionMap` is empty there is nothing to replace so
we should always return empty substitution type and/or conformance ref.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility
